P1684-mdarray/P1684.md

```diff
@@ -644,6 +651,482 @@ The larger or equal is intentional, because the mapping may already be not exhau
+for example with different extents or layout.
+In the example below, functions `f0` and `f1`
+interpret the storage represented by a `std::vector<float>`
+in different ways.  Neither of the two functions allocate,
+and each passes along the `std::vector<float>` to the next operation.
+The hypothetical `extract_container()` member function
+of `mdarray` would move the container out of the `mdarray`
+and return the moved-from container as `mdarray::container_type&&`.
+
+```c++
+std::vector<float>&& f0(std::vector<float>&& storage)
+{
+  mdarray m0{storage, 2, 3, 4};
+  // ... compute with m0 ...
+  return m0.extract_container();
+}
+
+std::vector<float>&& f1(std::vector<float>&& storage)
+{
+  mdarray m1{storage, 6, 4};
+  // ... compute with m1 ...
+  return m1.extract_container();
+}
+
+int main()
+{
+  // Only one vector allocation, created as a temporary,
+  // and moved into used_storage.
+  std::vector<float> used_storage(f1(f0(std::vector<float>(24, 0.0f))));
+  // ... use used_storage for other things ...
+  return 0;
+}
+```
+
+### Existing ways to get the same effects
+
+There are two reasons users may want to extract the container from an `mdarray`.
+
+1. They want to reuse the container's *storage*,
+    but not the *elements*.
+
+2. They want to reuse both *storage* and *elements*.
+
+In this section, we will explain how both of these use cases
+are covered by the current `mdarray` design,
+plus existing C++ Standard Library features.
+
+#### Reusing storage, but not elements
+
+Users may want to reuse the container's *storage*,
+but not the *elements*.
+That is, they do not care what the elements are
+when they create a new `mdarray`
+from a previously extracted container.
+
+This use case treats the container like a memory pool.
+That strongly suggests passing along a memory resource or allocator,
+instead of passing along a container.
+For example, they could use
+[`std::pmr::polymorphic_allocator`](http://en.cppreference.com/w/cpp/memory/polymorphic_allocator)
+with
+[`std::pmr::monotonic_buffer_resource`](https://en.cppreference.com/w/cpp/memory/monotonic_buffer_resource).
+The following shows how that might change
+the above `f0`, `f1` example.
+
+```c++
+pmr::monotonic_buffer_resource*
+f0(pmr::monotonic_buffer_resource* resource)
+{
+  pmr::polymorphic_allocator<float> allocator{resource};
+  mdarray m0{pmr::vector<float>{24, allocator}, 2, 3, 4};
+  // ... compute with m0 ...
+  return resource;
+}
+
+pmr::monotonic_buffer_resource*
+f1(pmr::monotonic_buffer_resource* resource)
+{
+  pmr::polymorphic_allocator<float> allocator{resource};
+  mdarray m1{pmr::vector<float>{24, allocator}, 6, 4};
+  // ... compute with m1 ...
+  return resource;
+}
+
+int main()
+{
+  vector<float> storage{24, 0.0f};
+  pmr::monotonic_buffer_resource mbr{
+    storage.data(), sizeof(float) * storage.size()};
+  pmr::monotonic_buffer_resource* mbr_ptr = f1(f0(mbr));
+  // ... use mbr_ptr for other things ... 
+  return 0;
+}
```

Comment:
I have virtually no experience with polymorphic allocators but I don't think your code snippet does what you think.
It does not reuse the storage.  See https://godbolt.org/z/7accYa79z

Reply:
Thanks for pointing this out!  I didn't actually test this example.  I probably should have ; - ) .  It's been a couple years since I had to do anything with `std::pmr`.

Reply:
I've fixed this example; thanks!